packages/codec/README.md

```diff
@@ -1,15 +1,161 @@
+to decode state variables will result in an exception if full mode is not
+possible.
+
+### Output format: `Type`s and `Result`s
+
+This will be a brief explanation of the format of decoded variables.  For full
+detail, I recommend seeing the documentation for `Format`.
+
+Each decoded value is a `Result`.  A `Result` has the following fields:
+
+1. `type`: This is a `Type` object describing the value's type.  Each `Type`
+has a `typeClass` field describing the overall broad type, such as `"uint"` or
+`"bytes"`, together with additional information that gives the specific type.  I
+won't go into further detail about these here; I recommend simply looking at
+`Format.Types` to see how these work.
+
+2. `kind`: This is either `"value"`, in which case the `Result` is a `Value`, or
+`"error"`, in which case the `Result` is an `ErrorResult`.  In the former case,
+there will be a `value` field containing the decoded value.  In the latter case,
+there will be an `error` field indicating what went wrong.  *Warning*: When
+decoding a complex type, such as an array, mapping, or array, getting a kind of
+`"value"` does not necessarily mean the individual elements were decoded
+successfully.  Even if the `Result` for the array (mapping, struct) as a whole
+has kind "value"`, the elements might still have kind `"error"`.
+
+3. `value`: As mentioned, this is included when `kind` is equal to `"value"`.
+It contains information about the actual decoded value.  I recommend seeing
+`Format.Values` for more information.
+
+4. `error`: The alternative to `value`.  Generally includes information about
+the raw data that led to the error.  See `Format.Errors` for more
+information.
+
+5. `reference`: This field is a debugger-only feature so I'll skip explaining it
+here; it won't come up when using this interface.
+ 
+Rather than explain the format in detail -- you can see the TypeDoc for that -- I'd
+instead like to take a moment to answer the question: What counts as a value,
+and what counts as an error?
+
+In general, the answer is that anything that can be generated via Solidity
+alone (i.e. no assembly), with correctly-encoded inputs, and without making use
+of compiler bugs, is a value, not an error.  That means that, for instance, the
+following things are values, not errors:
+
+1. A variable of contract type whose address does not actually hold a contract
+of that type;
+
+2. An external function pointer that does not correspond to a valid function;
+
+3. A string containing invalid UTF-8;
+
+etc.
+
+By contrast, the following *are* errors:
+
+1. A `bool` which is neither `false` (0) nor `true` (1);
+
+2. An `enum` which is out of range;
+
+etc.
+
+(You may be wondering about the enum case here, because if you go sufficiently
+far back, to Solidity 0.4.4 or earlier, it *was* possible to generate
+out-of-range enums without resorting to assembly or compiler bugs.  However, the
+decoder is not intended to support Solidity older than 0.4.9 (except in ABI-only
```

Comment:
It might be worth putting something closer to the top that says what Solidity versions we support

Reply:
Probably the place to put it would be the discussion of decoding modes -- mention that full mode is only supported for 0.4.9 and later.

Reply:
That makes sense; I'm just looking from the perspective that we will have three kinds of readers:
1. People that just want to get started quickly using the decoder
2. People that want to read the README first
3. People that have previously read the README and need to remember some detail

I am specifically thinking of (3) here.